File: backend/java/localpos-user-service/src/main/java/com/pds/localpos/userservice/model/Role.java
package com.pds.localpos.userservice.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "roles", schema = "users", indexes = {
        @Index(name = "idx_roles_name", columnList = "name")
})
public class Role {

    @Id
    private String id;

    @Column(nullable = false, unique = true, length = 50)
    private String name;

    @Column(length = 255)
    private String description;

    @ManyToMany(mappedBy = "roles", fetch = FetchType.LAZY) // o EAGER si prefieres
    @Builder.Default
    private Set<User> users = new HashSet<>();

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || org.hibernate.Hibernate.getClass(this) != org.hibernate.Hibernate.getClass(o)) {
            return false;
        }
        Role role = (Role) o;
        return id != null && id.equals(role.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        String desc = description == null ? null :
                description.length() > 50 ? description.substring(0, 50) + "..." : description;
        return "Role(id=" + id + ", name=" + name + ", description=" + desc + ")";
    }
}

File: backend/kotlin/localpos-user-service-kt/src/main/kotlin/com/pds/localpos/userservice/model/Role.kt
package com.pds.localpos.userservice.model

import jakarta.persistence.*
import org.hibernate.Hibernate
import java.util.*

@Entity
@Table(
    name = "roles",
    schema = "users",
    indexes = [Index(name = "idx_roles_name", columnList = "name")]
)
class Role(

    @Id
    @Column(length = 36, updatable = false, nullable = false)
    var id: String = UUID.randomUUID().toString(),

    @Column(nullable = false, unique = true, length = 50)
    var name: String = "",

    @Column(length = 255)
    var description: String? = null,

    @ManyToMany(mappedBy = "roles", fetch = FetchType.LAZY)
    var users: MutableSet<User> = mutableSetOf()
) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || Hibernate.getClass(this) != Hibernate.getClass(other)){
            return false
        }
        other as Role
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()

    override fun toString(): String {
        val desc = description?.let { if (it.length > 50) it.substring(0, 50) + "..." else it }
        return "Role(id=$id, name=$name, description=$desc)"
    }
}
